src/maxwell-web-view.js

```diff
@@ -0,0 +1,242 @@
+
+        /* Collect new positions */
+        positions.push ({ id: id, x: x, y: y });
+    }
+
+    /* Update all positions in MaxwellWebView at once to reduce messages */
+    window.webkit.messageHandlers.maxwell_update_positions.postMessage(positions);
+    position_timeout_id = null;
+}
+
+/* Limit to 10Hz */
+let update_positions_throttled = throttle(update_positions, 100);
+
+/* We need to update widget positions on scroll and resize events */
+window.addEventListener("scroll", update_positions_throttled, { passive: true });
+window.addEventListener("resize", update_positions_throttled, { passive: true });
+
+/* We also need to update it on any DOM change */
+function document_mutation_handler (mutations) {
+    let children = new Array ();
+
+    for (var mutation of mutations) {
+        if (mutation.type !== 'childList')
+            continue;
+
+        for (let i = 0, len = mutation.addedNodes.length; i < len; i++) {
+            let child = mutation.addedNodes[i];
+
+            if (!child.id || child.tagName !== 'CANVAS' ||
+                !child.classList.contains('GtkWidget'))
+                continue;
+
+            /* Save original display value */
+            child.maxwell_display_value = child.style.display;
+
+            /* Hide all widgets by default */
+            child.style.display = 'none';
+
+            /* And set no size (canvas default is 300x150) */
+            child.width = 0;
+            child.height = 0;
+
+            /* Keep a reference in a hash table for quick access */
+            maxwell.children[child.id] = child;
+
+            /* Collect children to allocate */
+            children.push(child.id);
+        }
+    }
+
+    /* Allocate GtkWidget, canvas size will change the first time
+     * child_draw() is called and we actually have something to show
+     */
+    window.webkit.messageHandlers.maxwell_children_allocate.postMessage(children);
+
+    /* Extra paranoid, update positions if anything changes in the DOM tree!
+     * ideally it would be nice to directly observe BoundingClientRect changes.
+     */
+    update_positions_throttled();
+};
+
+/* Main DOM observer */
+let observer = new MutationObserver(document_mutation_handler);
+observer.observe(document, {
+    childList: true,
+    subtree: true,
+    attributes: true
+});
+
+/* Semi Public API */
+
+/* Main entry point */
+window.maxwell = {};
+
+/* List of children */
+window.maxwell.children = {};
+
+/* child_resize()
+ */
+window.maxwell.child_resize = function (id, width, height) {
+    let child = maxwell.children[id];
+
+    if (!child)
+        return;
+
+    /* Resize canvas keeping the old contents */
+    if (child.width !== width || child.height !== height) {
+        let ctx = child.getContext('2d');
+        var old_image = ctx.getImageData(0, 0, child.width, child.height);
+
+        child.width = width;
+        child.height = height;
+
+        ctx.putImageData(old_image, 0, 0);
```

Comment:
I'm not sure I fully understand the logic here; you're fetching a `child.width X child.height` snapshot of the children, but then paint the same image data in the same exact position.
If `child` got larger, that should have no effect. If `child` got smaller, would it not be much faster to simply clip to the new size? What am I missing?